src/tls_stream.rs

```diff
@@ -374,13 +386,57 @@ impl<S> TlsStream<S>
         self.initialize().map(|_| ())
     }
 
+    /// Returns the negotiated application protocol for this tls stream, if one exists
+    /// # Remarks
+    /// This is only valid if the tls handshake is completed, and will return None if called
+    /// before this point.
+    pub fn get_negotiated_application_protocol(&mut self) -> Option<Vec<u8>> {
+        if let State::Streaming { .. } = self.state {
+            unsafe {
+                let mut client_protos: sspi::SecPkgContext_ApplicationProtocol = mem::zeroed();
+                let client_protos_ptr = &mut client_protos as *mut sspi::SecPkgContext_ApplicationProtocol;
+                assert_eq!(
```

Comment:
This will need to be replaced with proper error handling, but I'm not sure what the possible errors are here, and how we should deal with them as MSDN doesn't really give much info on that.